slides/src/content/redux.md

```diff
@@ -0,0 +1,356 @@
+1. User requests often modify the same objects but server requests usually modify different ones
+1. Users expect to be notified about changes immediately: servers are more patient
+
+---
+
+## Similarities of Structure - Redux vs Backend Architecture
+
+| Function                          | Angular With Redux   | Server              |
+|-----------------------------------|----------------------|---------------------|
+| Data Storage                      | Store                | Database            |
+| Presentation View                 | `@Component`         | Client API          |
+| Logic to massage data for Storage | `@Effects`           | Services            |
+| Communication                     | Action / Observables | Function Invocation |
+
+---
+
+## Redux as a System
+
+<img src="/content/images/redux-and-data-flow.png" width="50%"/>
+
+---
+
+## The Store
+
+- Stores the entire current state of the application
+- In principle, entire state replaced each time something changes
+  - I.e., state is never modified in place
+  - So no need to worry about concurrency effects
+  - In practice, can often replace some parts and copy other parts forward
+- Application components watch [Observable](http://reactivex.io/rxjs/) streams for updates
+  and change what they display
+
+FIXME: diagram of state update
+
+---
+
+## Actions
+
+- Actions are objects that tell the store how to update itself
+  - Use actions instead of direct function invocation so that they can be serialized
+- Each action is an `Object` with:
+  - `type`: identifies what kind of action this is (to allow selection and filtering)
+  - `payload`: extra information needed to carry out the action
+    - Name is just a convention, but a widely-used one
+
+---
+
+## Plan for Refactoring
+
+1. Define the *reducer* that turns a state and an action into a new state
+1. Define actions
+1. Main component will dispatch an action to add an item
+1. Display will monitor the store for changes
+
+Notes:
+
+1. We will use [@ngrx](https://github.com/ngrx) for our store
+   - Many other implementations of the Redux pattern exist
+1. We will leave room in our store to add more state later
+1. Angular CLI doesn't know anything about @ngrx, so we have to do most of the work by hand
+
+---
+
+## Install Required Software
+
+- `npm install @ngrx/core @ngrx/store --save`
+- The `--save` option updates `package.json`
+- So the next person can just do `npm install`
+
+---
+
+## Create the Reducer
+
+- Create a new file `src/app/store.ts`
+- Import `Action` to define the shape of actions
+- Define constants for action names (as strings)
+- Also define the initial state
+  - An empty item list
+- New state is the old list plus a new item
+  - New item arrives as action's payload
+- Do *not* use `items.push` to update existing state
+  - More efficient…
+  - …but only if correctness and programmer time aren't issues
+
```

Comment:
I find the following diagram very useful - Maybe it can be included. 

https://camo.githubusercontent.com/9de527b9432cc9244dc600875b46b43311918b59/68747470733a2f2f73332e616d617a6f6e6177732e636f6d2f6d656469612d702e736c69642e65732f75706c6f6164732f3336343831322f696d616765732f323438343739302f415243482d5265647578322d657874656e6465642d7265616c2d6465636c657261746976652e676966